File: src/shairport.ts
import { ChildProcess, spawn } from 'child_process';
import { EventEmitter } from 'events';
import { Readable } from 'stream';
import { AppConfig } from './config';
import { TrackMetadata } from './types';

export class ShairportManager extends EventEmitter {
  private process: ChildProcess | null = null;
  private config: AppConfig;
  private restarting = false;
  private restartAttempts = 0;
  private maxRestartAttempts = 10;
  private restartDelay = 2000;

  constructor(config: AppConfig) {
    super();
    this.config = config;
  }

  async start(): Promise<void> {
    await this.validateBinary();
    this.spawnProcess();
  }

  private async validateBinary(): Promise<void> {
    return new Promise((resolve, reject) => {
      const check = spawn(this.config.shairportPath, ['--version'], {
        stdio: ['ignore', 'pipe', 'pipe'],
      });

      let output = '';
      check.stdout?.on('data', (d) => (output += d.toString()));
      check.stderr?.on('data', (d) => (output += d.toString()));

      check.on('close', (code) => {
        // shairport-sync --version may exit with 0 or 1 depending on version
        if (output.toLowerCase().includes('shairport')) {
          console.log(`[shairport] Found: ${output.trim()}`);
          resolve();
        } else if (code === 0) {
          resolve();
        } else {
          reject(
            new Error(
              `shairport-sync binary not found or not working at "${this.config.shairportPath}". ` +
                `Install it with: sudo apt install shairport-sync (Linux) or brew install shairport-sync (macOS). ` +
                `Output: ${output}`
            )
          );
        }
      });

      check.on('error', (err) => {
        reject(
          new Error(
            `shairport-sync binary not found at "${this.config.shairportPath}". ` +
              `Install it with: sudo apt install shairport-sync (Linux) or brew install shairport-sync (macOS). ` +
              `Error: ${err.message}`
          )
        );
      });
    });
  }

  private spawnProcess(): void {
    const args = [
      '--name', this.config.receiverName,
      '--output', 'stdout',
      // Output raw S16LE PCM at configured sample rate
      '-v', // verbose mode for metadata on stderr
    ];

    console.log(`[shairport] Starting: ${this.config.shairportPath} ${args.join(' ')}`);

    this.process = spawn(this.config.shairportPath, args, {
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    if (this.process.stdout) {
      this.restartAttempts = 0;
      this.emit('audio', this.process.stdout);
    }

    if (this.process.stderr) {
      let buffer = '';
      this.process.stderr.on('data', (data: Buffer) => {
        buffer += data.toString();
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';
        for (const line of lines) {
          this.parseMetadataLine(line);
        }
      });
    }

    this.process.on('close', (code, signal) => {
      console.log(`[shairport] Process exited (code=${code}, signal=${signal})`);
      this.process = null;
      this.emit('stopped');

      if (!this.restarting && code !== 0 && signal !== 'SIGTERM') {
        this.scheduleRestart();
      }
    });

    this.process.on('error', (err) => {
      console.error(`[shairport] Process error: ${err.message}`);
      this.emit('error', err);
    });

    this.emit('started');
  }

  private parseMetadataLine(line: string): void {
    // shairport-sync verbose output includes metadata hints
    const metadata: TrackMetadata = {};

    if (line.includes('Title:')) {
      metadata.title = line.split('Title:')[1]?.trim();
    } else if (line.includes('Artist:')) {
      metadata.artist = line.split('Artist:')[1]?.trim();
    } else if (line.includes('Album:')) {
      metadata.album = line.split('Album:')[1]?.trim();
    }

    if (metadata.title || metadata.artist || metadata.album) {
      this.emit('metadata', metadata);
    }
  }

  private scheduleRestart(): void {
    if (this.restartAttempts >= this.maxRestartAttempts) {
      console.error(`[shairport] Max restart attempts (${this.maxRestartAttempts}) reached. Giving up.`);
      this.emit('error', new Error('shairport-sync crashed too many times'));
      return;
    }

    this.restartAttempts++;
    const delay = this.restartDelay * Math.min(this.restartAttempts, 5);
    console.log(`[shairport] Restarting in ${delay}ms (attempt ${this.restartAttempts}/${this.maxRestartAttempts})`);

    setTimeout(() => {
      if (!this.process) {
        this.spawnProcess();
      }
    }, delay);
  }

  async stop(): Promise<void> {
    this.restarting = true;
    if (this.process) {
      this.process.kill('SIGTERM');
      await new Promise<void>((resolve) => {
        const timeout = setTimeout(() => {
          if (this.process) {
            this.process.kill('SIGKILL');
          }
          resolve();
        }, 5000);

        if (this.process) {
          this.process.on('close', () => {
            clearTimeout(timeout);
            resolve();
          });
        } else {
          clearTimeout(timeout);
          resolve();
        }
      });
      this.process = null;
    }
    this.restarting = false;
  }

  isRunning(): boolean {
    return this.process !== null;
  }
}

File: src/shairport.rs
use anyhow::{anyhow, Result};
use bytes::Bytes;
use tokio::io::{AsyncBufReadExt, AsyncReadExt, BufReader};
use tokio::process::{Child, Command};
use tokio::sync::broadcast;

use crate::types::TrackMetadata;

pub enum ShairportEvent {
    Started,
    Stopped,
    Metadata(TrackMetadata),
    Error(String),
}

pub struct ShairportManager {
    shairport_path: String,
    receiver_name: String,
    child: Option<Child>,
}

impl ShairportManager {
    pub fn new(shairport_path: &str, receiver_name: &str) -> Self {
        Self {
            shairport_path: shairport_path.to_string(),
            receiver_name: receiver_name.to_string(),
            child: None,
        }
    }

    pub async fn validate_binary(&self) -> Result<()> {
        let output = Command::new(&self.shairport_path)
            .arg("--version")
            .output()
            .await
            .map_err(|e| {
                anyhow!(
                    "shairport-sync binary not found at \"{}\". \
                     Install: sudo apt install shairport-sync (Linux) or \
                     brew install shairport-sync (macOS). Error: {}",
                    self.shairport_path,
                    e
                )
            })?;

        let combined = String::from_utf8_lossy(&output.stdout).to_string()
            + &String::from_utf8_lossy(&output.stderr);

        if combined.to_lowercase().contains("shairport") || output.status.success() {
            tracing::info!("[shairport] Found: {}", combined.trim());
            Ok(())
        } else {
            Err(anyhow!("shairport-sync not working. Output: {}", combined))
        }
    }

    pub async fn start(
        &mut self,
        audio_tx: broadcast::Sender<Bytes>,
        event_tx: tokio::sync::mpsc::Sender<ShairportEvent>,
    ) -> Result<()> {
        self.validate_binary().await?;
        self.spawn_process(audio_tx, event_tx).await
    }

    async fn spawn_process(
        &mut self,
        audio_tx: broadcast::Sender<Bytes>,
        event_tx: tokio::sync::mpsc::Sender<ShairportEvent>,
    ) -> Result<()> {
        let mut child = Command::new(&self.shairport_path)
            .args([
                "--name",
                &self.receiver_name,
                "--output",
                "stdout",
                "-v",
            ])
            .stdin(std::process::Stdio::null())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .spawn()?;

        let stdout = child.stdout.take().expect("stdout piped");
        let stderr = child.stderr.take().expect("stderr piped");

        // Read stdout PCM data and publish to broadcast channel
        tokio::spawn(async move {
            let mut reader = BufReader::new(stdout);
            let mut buf = vec![0u8; 4096];
            loop {
                match reader.read(&mut buf).await {
                    Ok(0) => break,
                    Ok(n) => {
                        let chunk = Bytes::copy_from_slice(&buf[..n]);
                        let _ = audio_tx.send(chunk);
                    }
                    Err(e) => {
                        tracing::error!("[shairport] stdout read error: {}", e);
                        break;
                    }
                }
            }
        });

        // Read stderr for metadata lines
        let event_tx_clone = event_tx.clone();
        tokio::spawn(async move {
            let reader = BufReader::new(stderr);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                if let Some(meta) = parse_metadata_line(&line) {
                    let _ = event_tx_clone.send(ShairportEvent::Metadata(meta)).await;
                }
            }
        });

        let _ = event_tx.send(ShairportEvent::Started).await;
        self.child = Some(child);
        Ok(())
    }

    pub async fn stop(&mut self) {
        if let Some(mut child) = self.child.take() {
            let _ = child.kill().await;
            let _ = child.wait().await;
        }
    }

    pub fn is_running(&self) -> bool {
        self.child.is_some()
    }
}

pub fn parse_metadata_line(line: &str) -> Option<TrackMetadata> {
    let mut meta = TrackMetadata::default();

    if line.contains("Title:") {
        meta.title = line.split("Title:").nth(1).map(|s| s.trim().to_string());
    } else if line.contains("Artist:") {
        meta.artist = line.split("Artist:").nth(1).map(|s| s.trim().to_string());
    } else if line.contains("Album:") {
        meta.album = line.split("Album:").nth(1).map(|s| s.trim().to_string());
    }

    if meta.title.is_some() || meta.artist.is_some() || meta.album.is_some() {
        Some(meta)
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_title() {
        let meta = parse_metadata_line("  Title: My Song").unwrap();
        assert_eq!(meta.title.unwrap(), "My Song");
        assert!(meta.artist.is_none());
        assert!(meta.album.is_none());
    }

    #[test]
    fn test_parse_artist() {
        let meta = parse_metadata_line("  Artist: The Band").unwrap();
        assert!(meta.title.is_none());
        assert_eq!(meta.artist.unwrap(), "The Band");
    }

    #[test]
    fn test_parse_album() {
        let meta = parse_metadata_line("  Album: Greatest Hits").unwrap();
        assert_eq!(meta.album.unwrap(), "Greatest Hits");
    }

    #[test]
    fn test_parse_non_metadata_line() {
        assert!(parse_metadata_line("Playing audio...").is_none());
        assert!(parse_metadata_line("").is_none());
    }
}
